File: sdk/golang/cmd/prismer/security.go
package main

import (
	"fmt"

	prismer "github.com/Prismer-AI/PrismerCloud/sdk/golang"
	"github.com/spf13/cobra"
)

// ============================================================================
// security command group
// ============================================================================

var securityCmd = &cobra.Command{
	Use:   "security",
	Short: "Per-conversation encryption and key management",
}

// ── security get ────────────────────────────────────────

var securityGetJSON bool
var securityGetCmd = &cobra.Command{
	Use:   "get <conversation-id>",
	Short: "Get security settings for a conversation",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Security.GetConversationSecurity(ctx, args[0])
		if err != nil {
			return err
		}
		if securityGetJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		data := asMap(res.Data)
		fmt.Printf("Encryption Mode: %v\n", data["encryptionMode"])
		fmt.Printf("Signing Policy:  %v\n", data["signingPolicy"])
		return nil
	},
}

// ── security set ────────────────────────────────────────

var securitySetMode string
var securitySetJSON bool
var securitySetCmd = &cobra.Command{
	Use:   "set <conversation-id>",
	Short: "Set encryption mode for a conversation",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Security.SetConversationSecurity(ctx, args[0], map[string]interface{}{
			"encryptionMode": securitySetMode,
		})
		if err != nil {
			return err
		}
		if securitySetJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		fmt.Printf("Encryption mode set to: %s\n", securitySetMode)
		return nil
	},
}

// ── security upload-key ─────────────────────────────────

var (
	securityUploadKeyKey       string
	securityUploadKeyAlgorithm string
	securityUploadKeyJSON      bool
)

var securityUploadKeyCmd = &cobra.Command{
	Use:   "upload-key <conversation-id>",
	Short: "Upload an ECDH public key for a conversation",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Security.UploadKey(ctx, args[0], securityUploadKeyKey, securityUploadKeyAlgorithm)
		if err != nil {
			return err
		}
		if securityUploadKeyJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		fmt.Printf("Key uploaded (algorithm: %s)\n", securityUploadKeyAlgorithm)
		return nil
	},
}

// ── security keys ───────────────────────────────────────

var securityKeysJSON bool
var securityKeysCmd = &cobra.Command{
	Use:   "keys <conversation-id>",
	Short: "List all member public keys for a conversation",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Security.GetKeys(ctx, args[0])
		if err != nil {
			return err
		}
		if securityKeysJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		keys := asList(res.Data)
		if len(keys) == 0 {
			fmt.Println("No keys found.")
			return nil
		}
		fmt.Printf("%-36s  %-16s  %s\n", "User ID", "Algorithm", "Public Key")
		for _, k := range keys {
			km := asMap(k)
			fmt.Printf("%-36v  %-16v  %v\n", km["userId"], km["algorithm"], km["publicKey"])
		}
		return nil
	},
}

// ── security revoke-key ─────────────────────────────────

var securityRevokeKeyJSON bool
var securityRevokeKeyCmd = &cobra.Command{
	Use:   "revoke-key <conversation-id> <user-id>",
	Short: "Revoke a member key from a conversation",
	Args:  cobra.ExactArgs(2),
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Security.RevokeKey(ctx, args[0], args[1])
		if err != nil {
			return err
		}
		if securityRevokeKeyJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		fmt.Printf("Key revoked for user: %s\n", args[1])
		return nil
	},
}

// ============================================================================
// identity command group
// ============================================================================

var identityCmd = &cobra.Command{
	Use:   "identity",
	Short: "Identity key management and audit log verification",
}

// ── identity server-key ─────────────────────────────────

var identityServerKeyJSON bool
var identityServerKeyCmd = &cobra.Command{
	Use:   "server-key",
	Short: "Get the server's identity public key",
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Identity.GetServerKey(ctx)
		if err != nil {
			return err
		}
		if identityServerKeyJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		data := asMap(res.Data)
		fmt.Printf("Server Public Key: %v\n", data["publicKey"])
		return nil
	},
}

// ── identity register-key ───────────────────────────────

var (
	identityRegisterKeyAlgorithm string
	identityRegisterKeyPublicKey string
	identityRegisterKeyJSON      bool
)

var identityRegisterKeyCmd = &cobra.Command{
	Use:   "register-key",
	Short: "Register an identity public key",
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Identity.RegisterKey(ctx, &prismer.RegisterKeyOptions{
			PublicKey: identityRegisterKeyPublicKey,
		})
		if err != nil {
			return err
		}
		if identityRegisterKeyJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		fmt.Printf("Identity key registered (algorithm: %s)\n", identityRegisterKeyAlgorithm)
		return nil
	},
}

// ── identity get-key ────────────────────────────────────

var identityGetKeyJSON bool
var identityGetKeyCmd = &cobra.Command{
	Use:   "get-key <user-id>",
	Short: "Get a user's identity public key",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Identity.GetKey(ctx, args[0])
		if err != nil {
			return err
		}
		if identityGetKeyJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		data := asMap(res.Data)
		fmt.Printf("Algorithm:  %v\n", data["algorithm"])
		fmt.Printf("Public Key: %v\n", data["publicKey"])
		return nil
	},
}

// ── identity revoke-key ─────────────────────────────────

var identityRevokeKeyJSON bool
var identityRevokeKeyCmd = &cobra.Command{
	Use:   "revoke-key",
	Short: "Revoke your own identity key",
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Identity.RevokeKey(ctx)
		if err != nil {
			return err
		}
		if identityRevokeKeyJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		fmt.Println("Identity key revoked.")
		return nil
	},
}

// ── identity audit-log ──────────────────────────────────

var identityAuditLogJSON bool
var identityAuditLogCmd = &cobra.Command{
	Use:   "audit-log <user-id>",
	Short: "Get key audit log entries for a user",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Identity.GetAuditLog(ctx, args[0])
		if err != nil {
			return err
		}
		if identityAuditLogJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		entries := asList(res.Data)
		if len(entries) == 0 {
			fmt.Println("No audit log entries.")
			return nil
		}
		fmt.Printf("%-24s  %-20s  %s\n", "Date", "Action", "Details")
		for _, e := range entries {
			em := asMap(e)
			fmt.Printf("%-24v  %-20v  %v\n", em["createdAt"], em["action"], em["details"])
		}
		return nil
	},
}

// ── identity verify-audit ───────────────────────────────

var identityVerifyAuditJSON bool
var identityVerifyAuditCmd = &cobra.Command{
	Use:   "verify-audit <user-id>",
	Short: "Verify the integrity of the key audit log for a user",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Identity.VerifyAuditLog(ctx, args[0])
		if err != nil {
			return err
		}
		if identityVerifyAuditJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		data := asMap(res.Data)
		if valid, ok := data["valid"].(bool); ok && valid {
			fmt.Println("Audit log verified: VALID")
		} else {
			fmt.Println("Audit log verified: INVALID")
		}
		return nil
	},
}

func init() {
	// security
	securitySetCmd.Flags().StringVar(&securitySetMode, "mode", "", "Encryption mode: none, available, or required")
	securitySetCmd.Flags().BoolVar(&securitySetJSON, "json", false, "Output raw JSON")
	_ = securitySetCmd.MarkFlagRequired("mode")
	securityGetCmd.Flags().BoolVar(&securityGetJSON, "json", false, "Output raw JSON")
	securityUploadKeyCmd.Flags().StringVar(&securityUploadKeyKey, "key", "", "Base64-encoded public key")
	securityUploadKeyCmd.Flags().StringVar(&securityUploadKeyAlgorithm, "algorithm", "ecdh-p256", "Key algorithm")
	securityUploadKeyCmd.Flags().BoolVar(&securityUploadKeyJSON, "json", false, "Output raw JSON")
	_ = securityUploadKeyCmd.MarkFlagRequired("key")
	securityKeysCmd.Flags().BoolVar(&securityKeysJSON, "json", false, "Output raw JSON")
	securityRevokeKeyCmd.Flags().BoolVar(&securityRevokeKeyJSON, "json", false, "Output raw JSON")

	securityCmd.AddCommand(securityGetCmd)
	securityCmd.AddCommand(securitySetCmd)
	securityCmd.AddCommand(securityUploadKeyCmd)
	securityCmd.AddCommand(securityKeysCmd)
	securityCmd.AddCommand(securityRevokeKeyCmd)
	rootCmd.AddCommand(securityCmd)

	// identity
	identityServerKeyCmd.Flags().BoolVar(&identityServerKeyJSON, "json", false, "Output raw JSON")
	identityRegisterKeyCmd.Flags().StringVar(&identityRegisterKeyAlgorithm, "algorithm", "", "Key algorithm (e.g. ed25519, ecdh-p256)")
	identityRegisterKeyCmd.Flags().StringVar(&identityRegisterKeyPublicKey, "public-key", "", "Base64-encoded public key")
	identityRegisterKeyCmd.Flags().BoolVar(&identityRegisterKeyJSON, "json", false, "Output raw JSON")
	_ = identityRegisterKeyCmd.MarkFlagRequired("algorithm")
	_ = identityRegisterKeyCmd.MarkFlagRequired("public-key")
	identityGetKeyCmd.Flags().BoolVar(&identityGetKeyJSON, "json", false, "Output raw JSON")
	identityRevokeKeyCmd.Flags().BoolVar(&identityRevokeKeyJSON, "json", false, "Output raw JSON")
	identityAuditLogCmd.Flags().BoolVar(&identityAuditLogJSON, "json", false, "Output raw JSON")
	identityVerifyAuditCmd.Flags().BoolVar(&identityVerifyAuditJSON, "json", false, "Output raw JSON")

	identityCmd.AddCommand(identityServerKeyCmd)
	identityCmd.AddCommand(identityRegisterKeyCmd)
	identityCmd.AddCommand(identityGetKeyCmd)
	identityCmd.AddCommand(identityRevokeKeyCmd)
	identityCmd.AddCommand(identityAuditLogCmd)
	identityCmd.AddCommand(identityVerifyAuditCmd)
	rootCmd.AddCommand(identityCmd)
}

File: sdk/prismer-cloud/typescript/src/commands/security.ts
import { Command } from 'commander';
import { PrismerClient } from '../index';

type ClientFactory = () => PrismerClient;

export function register(parent: Command, getIMClient: ClientFactory, _getAPIClient: ClientFactory): void {
  // ---------------------------------------------------------------------------
  // security command group
  // ---------------------------------------------------------------------------
  const security = parent
    .command('security')
    .description('Per-conversation encryption and key management');

  // security get <conversation-id>
  security
    .command('get <conversation-id>')
    .description('Get security settings for a conversation')
    .option('--json', 'Output raw JSON response')
    .action(async (convId: string, opts: { json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.security.getConversationSecurity(convId);
        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || JSON.stringify(res.error)}\n`);
          process.exit(1);
        }
        if (opts.json) {
          process.stdout.write(JSON.stringify(res.data, null, 2) + '\n');
          return;
        }
        const d = res.data as Record<string, unknown> | undefined;
        process.stdout.write(`Encryption Mode: ${d?.encryptionMode ?? '-'}\n`);
        process.stdout.write(`Signing Policy:  ${d?.signingPolicy ?? '-'}\n`);
      } catch (err: unknown) {
        process.stderr.write(`Error: ${err instanceof Error ? err.message : String(err)}\n`);
        process.exit(1);
      }
    });

  // security set <conversation-id>
  security
    .command('set <conversation-id>')
    .description('Set encryption mode for a conversation')
    .requiredOption('--mode <mode>', 'Encryption mode: none, available, or required')
    .option('--json', 'Output raw JSON response')
    .action(async (convId: string, opts: { mode: string; json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.security.setConversationSecurity(convId, {
          encryptionMode: opts.mode as 'none' | 'available' | 'required',
        });
        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || JSON.stringify(res.error)}\n`);
          process.exit(1);
        }
        if (opts.json) {
          process.stdout.write(JSON.stringify(res.data, null, 2) + '\n');
          return;
        }
        process.stdout.write(`Encryption mode set to: ${opts.mode}\n`);
      } catch (err: unknown) {
        process.stderr.write(`Error: ${err instanceof Error ? err.message : String(err)}\n`);
        process.exit(1);
      }
    });

  // security upload-key <conversation-id>
  security
    .command('upload-key <conversation-id>')
    .description('Upload an ECDH public key for a conversation')
    .requiredOption('--key <base64>', 'Base64-encoded public key')
    .option('--algorithm <alg>', 'Key algorithm', 'ecdh-p256')
    .option('--json', 'Output raw JSON response')
    .action(async (convId: string, opts: { key: string; algorithm: string; json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.security.uploadKey(convId, opts.key, opts.algorithm);
        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || JSON.stringify(res.error)}\n`);
          process.exit(1);
        }
        if (opts.json) {
          process.stdout.write(JSON.stringify(res.data, null, 2) + '\n');
          return;
        }
        process.stdout.write(`Key uploaded (algorithm: ${opts.algorithm})\n`);
      } catch (err: unknown) {
        process.stderr.write(`Error: ${err instanceof Error ? err.message : String(err)}\n`);
        process.exit(1);
      }
    });

  // security keys <conversation-id>
  security
    .command('keys <conversation-id>')
    .description('List all member public keys for a conversation')
    .option('--json', 'Output raw JSON response')
    .action(async (convId: string, opts: { json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.security.getKeys(convId);
        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || JSON.stringify(res.error)}\n`);
          process.exit(1);
        }
        const keys = res.data as unknown[] | undefined;
        if (opts.json) {
          process.stdout.write(JSON.stringify(keys, null, 2) + '\n');
          return;
        }
        if (!keys || (Array.isArray(keys) && keys.length === 0)) {
          process.stdout.write('No keys found.\n');
          return;
        }
        process.stdout.write('User ID'.padEnd(36) + 'Algorithm'.padEnd(16) + 'Public Key\n');
        for (const k of keys as Array<Record<string, unknown>>) {
          process.stdout.write(
            `${String(k.userId ?? '').padEnd(36)}${String(k.algorithm ?? '').padEnd(16)}${String(k.publicKey ?? '')}\n`,
          );
        }
      } catch (err: unknown) {
        process.stderr.write(`Error: ${err instanceof Error ? err.message : String(err)}\n`);
        process.exit(1);
      }
    });

  // security revoke-key <conversation-id> <user-id>
  security
    .command('revoke-key <conversation-id> <user-id>')
    .description('Revoke a member key from a conversation')
    .option('--json', 'Output raw JSON response')
    .action(async (convId: string, userId: string, opts: { json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.security.revokeKey(convId, userId);
        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || JSON.stringify(res.error)}\n`);
          process.exit(1);
        }
        if (opts.json) {
          process.stdout.write(JSON.stringify(res.data, null, 2) + '\n');
          return;
        }
        process.stdout.write(`Key revoked for user: ${userId}\n`);
      } catch (err: unknown) {
        process.stderr.write(`Error: ${err instanceof Error ? err.message : String(err)}\n`);
        process.exit(1);
      }
    });

  // ---------------------------------------------------------------------------
  // identity command group
  // ---------------------------------------------------------------------------
  const identity = parent
    .command('identity')
    .description('Identity key management and audit log verification');

  // identity server-key
  identity
    .command('server-key')
    .description("Get the server's identity public key")
    .option('--json', 'Output raw JSON response')
    .action(async (opts: { json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.identity.getServerKey();
        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || JSON.stringify(res.error)}\n`);
          process.exit(1);
        }
        if (opts.json) {
          process.stdout.write(JSON.stringify(res.data, null, 2) + '\n');
          return;
        }
        const d = res.data as { publicKey?: string } | undefined;
        process.stdout.write(`Server Public Key: ${d?.publicKey ?? '-'}\n`);
      } catch (err: unknown) {
        process.stderr.write(`Error: ${err instanceof Error ? err.message : String(err)}\n`);
        process.exit(1);
      }
    });

  // identity register-key
  identity
    .command('register-key')
    .description('Register an identity public key')
    .requiredOption('--algorithm <alg>', 'Key algorithm (e.g. ed25519, ecdh-p256)')
    .requiredOption('--public-key <base64>', 'Base64-encoded public key')
    .option('--json', 'Output raw JSON response')
    .action(async (opts: { algorithm: string; publicKey: string; json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.identity.registerKey({
          algorithm: opts.algorithm,
          publicKey: opts.publicKey,
        });
        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || JSON.stringify(res.error)}\n`);
          process.exit(1);
        }
        if (opts.json) {
          process.stdout.write(JSON.stringify(res.data, null, 2) + '\n');
          return;
        }
        process.stdout.write(`Identity key registered (algorithm: ${opts.algorithm})\n`);
      } catch (err: unknown) {
        process.stderr.write(`Error: ${err instanceof Error ? err.message : String(err)}\n`);
        process.exit(1);
      }
    });

  // identity get-key <user-id>
  identity
    .command('get-key <user-id>')
    .description("Get a user's identity public key")
    .option('--json', 'Output raw JSON response')
    .action(async (userId: string, opts: { json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.identity.getKey(userId);
        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || JSON.stringify(res.error)}\n`);
          process.exit(1);
        }
        if (opts.json) {
          process.stdout.write(JSON.stringify(res.data, null, 2) + '\n');
          return;
        }
        const d = res.data as Record<string, unknown> | undefined;
        process.stdout.write(`Algorithm:  ${d?.algorithm ?? '-'}\n`);
        process.stdout.write(`Public Key: ${d?.publicKey ?? '-'}\n`);
      } catch (err: unknown) {
        process.stderr.write(`Error: ${err instanceof Error ? err.message : String(err)}\n`);
        process.exit(1);
      }
    });

  // identity revoke-key
  identity
    .command('revoke-key')
    .description('Revoke your own identity key')
    .option('--json', 'Output raw JSON response')
    .action(async (opts: { json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.identity.revokeKey();
        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || JSON.stringify(res.error)}\n`);
          process.exit(1);
        }
        if (opts.json) {
          process.stdout.write(JSON.stringify(res.data, null, 2) + '\n');
          return;
        }
        process.stdout.write('Identity key revoked.\n');
      } catch (err: unknown) {
        process.stderr.write(`Error: ${err instanceof Error ? err.message : String(err)}\n`);
        process.exit(1);
      }
    });

  // identity audit-log <user-id>
  identity
    .command('audit-log <user-id>')
    .description('Get key audit log entries for a user')
    .option('--json', 'Output raw JSON response')
    .action(async (userId: string, opts: { json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.identity.getAuditLog(userId);
        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || JSON.stringify(res.error)}\n`);
          process.exit(1);
        }
        const entries = res.data as unknown[] | undefined;
        if (opts.json) {
          process.stdout.write(JSON.stringify(entries, null, 2) + '\n');
          return;
        }
        if (!entries || (Array.isArray(entries) && entries.length === 0)) {
          process.stdout.write('No audit log entries.\n');
          return;
        }
        process.stdout.write('Date'.padEnd(24) + 'Action'.padEnd(20) + 'Details\n');
        for (const e of entries as Array<Record<string, unknown>>) {
          const date = e.createdAt ? new Date(String(e.createdAt)).toLocaleString() : '';
          process.stdout.write(
            `${date.padEnd(24)}${String(e.action ?? '').padEnd(20)}${e.details ? JSON.stringify(e.details) : ''}\n`,
          );
        }
      } catch (err: unknown) {
        process.stderr.write(`Error: ${err instanceof Error ? err.message : String(err)}\n`);
        process.exit(1);
      }
    });

  // identity verify-audit <user-id>
  identity
    .command('verify-audit <user-id>')
    .description('Verify the integrity of the key audit log for a user')
    .option('--json', 'Output raw JSON response')
    .action(async (userId: string, opts: { json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.identity.verifyAuditLog(userId);
        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || JSON.stringify(res.error)}\n`);
          process.exit(1);
        }
        if (opts.json) {
          process.stdout.write(JSON.stringify(res.data, null, 2) + '\n');
          return;
        }
        const d = res.data as { valid?: boolean; errors?: unknown[] } | undefined;
        if (d?.valid) {
          process.stdout.write('Audit log verified: VALID\n');
        } else {
          process.stdout.write('Audit log verified: INVALID\n');
          if (d?.errors && Array.isArray(d.errors) && d.errors.length > 0) {
            process.stdout.write('Errors:\n');
            for (const err of d.errors) {
              process.stdout.write(`  - ${JSON.stringify(err)}\n`);
            }
          }
        }
      } catch (err: unknown) {
        process.stderr.write(`Error: ${err instanceof Error ? err.message : String(err)}\n`);
        process.exit(1);
      }
    });
}
